File: GoL/ViewController.m
//
//  ViewController.m
//  GoL
//
//  Created by Marcel Meyer on 27.04.14.
//
//

#import "ViewController.h"
#import "Game.h"
#import "Dimensions.h"
#import "GameLoop.h"
#import "GameStateInitializerStrategy.h"
#import "RandomStrategy.h"
#import "AllDeadStrategy.h"

@interface ViewController ()
@property (strong, nonatomic) IBOutlet UITextView *gameResult;
@property (strong, nonatomic) IBOutlet UIButton *pauseResumeButton;
@property (strong, nonatomic) Game *game;
@property (strong, nonatomic) GameLoop *gameLoop;
@property (nonatomic) BOOL doUpdateGameResult;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.doUpdateGameResult = YES;
    [self.pauseResumeButton setEnabled:NO];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)touchRandom:(UIButton *)sender {
    [self createRandomGame];
    [self start];
}

- (IBAction)touchCross:(UIButton *)sender {
    [self.game addFigure:[Figure createCross]];
}

- (IBAction)touchShip:(id)sender {
    [self.game addFigure:[Figure createLightweightSpaceship]];
}

- (IBAction)touchBlank:(UIButton *)sender {
    [self createAllDeadGame];
    [self start];
}

- (void)pause {
    [self.gameLoop stop];
    [self.pauseResumeButton setTag:0];
    [self.pauseResumeButton setTitle:@"Resume" forState:UIControlStateNormal];
}

- (void)resume {
    [self.pauseResumeButton setTag:1];
    [self.pauseResumeButton setTitle:@"Pause" forState:UIControlStateNormal];
    [self.gameLoop start];
}

- (IBAction)touchPauseResume:(UIButton *)sender {
    if (sender.tag == 1) {
        [self pause];
    }
    else if (sender.tag == 0) {
        [self resume];
    }
}

- (void)start {
    ViewController *controller = self;

    if (self.gameLoop) {
        self.gameLoop.afterTick = nil;
        NSLog(@"Kill from Controller");
        [self.gameLoop killWithCallback: ^(void) {
            self.gameLoop = [[GameLoop alloc] initWithGame:self.game andTimeInterval:0.1 andAfterTick: ^(void) {
                [controller updateGameResult];
            }];

            [self.pauseResumeButton setEnabled:YES];
            [self resume];
        }];
    }
    else {
        self.gameLoop = [[GameLoop alloc] initWithGame:self.game andTimeInterval:0.1 andAfterTick: ^(void) {
            [controller updateGameResult];
        }];

        [self.pauseResumeButton setEnabled:YES];
        [self resume];
    }
}

- (void)updateGameResult {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.gameResult setText:self.game.description];
    });
}

- (void)createRandomGame {
    id gameStateInitializerStrategy = [[RandomStrategy alloc] init];

    self.game = [self createGame:gameStateInitializerStrategy];
}

- (void)createAllDeadGame {
    id gameStateInitializerStrategy = [[AllDeadStrategy alloc] init];

    self.game = [self createGame:gameStateInitializerStrategy];
}

- (Game *)createGame:(id <GameStateInitializerStrategy> )gameStateInitializerStrategy {
    Dimensions *dimensions = [[Dimensions alloc] initWithHeight:30 andWidth:44];

    GameStateInitializer *gameStateInitializer = [[GameStateInitializer alloc] initWithDimensions:dimensions andGameStateInitializerStrategy:gameStateInitializerStrategy];

    Game *game = [[Game alloc] initWithDimensions:dimensions andGameStateInitializer:gameStateInitializer];

    return game;
}

@end

File: GoL/ViewController.swift
//
//  ViewController.swift
//  GoL
//
//  Created by Georg Meyer on 08.06.14.
//
//

import Foundation
import UIKit

class NewViewController : UIViewController {
    
    @IBOutlet var gameResult : UITextView
    @IBOutlet var pauseResumeButton : UIButton
    
    var game : Game!
    var gameLoop : GameLoop!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.pauseResumeButton.enabled = false
        
    }
    
    @IBAction func touchRandom(sender : UIButton) {
        self.createRandomGame()
        self.start()
    }

    @IBAction func touchCross(sender : UIButton) {
        
    }
    
    @IBAction func touchShip(sender : UIButton) {
        
    }
    
    @IBAction func touchBlank(sender : UIButton) {
        self.createAllDeadGame()
        self.start()
    }
    
    @IBAction func touchPauseResume(sender : UIButton){
        if(sender.tag == 1){
            self.pause()
        } else if(sender.tag == 0){
            self.resume()
        }
    }
    
    func pause() {
        self.gameLoop?.stop()
        self.pauseResumeButton.tag = 0
        self.pauseResumeButton.setTitle("Resume", forState: UIControlState.Normal)
    }
    
    func resume() {
        self.gameLoop?.start()
        self.pauseResumeButton.tag = 1
        self.pauseResumeButton.setTitle("Pause", forState: UIControlState.Normal)
    }
    
    func createRandomGame() {
        let gameStateInitializerStrategy = RandomStrategy()
        
        self.game = createGame(gameStateInitializerStrategy)
    }
    
    func createAllDeadGame() {
        let gameStateInitializerStrategy = AllDeadStrategy()
        
        self.game = createGame(gameStateInitializerStrategy)
    }
    
    func start() {
        let controller = self
        self.gameLoop = GameLoop(game: self.game!, timeInterval: 0.1, afterTick: {
                controller.updateGameResult()
            })
        self.pauseResumeButton.enabled = true
        self.resume()
    }
    
    func updateGameResult() {
        dispatch_async(dispatch_get_main_queue(), {
            self.gameResult.text = self.game.description
        })
    }
    
    func createGame(gameStateInitializerStrategy : GameStateInitializerStrategy) -> Game {
        let dimensions = Dimensions(width: 44, height: 30)
        let gameStateInitializer = GameStateInitializer(dimensions: dimensions, gameStateInitializerStrategy: gameStateInitializerStrategy)
        
        return Game(dimensions: dimensions, gameStateInitializer: gameStateInitializer)
    }
}